surface/include/pcl/surface/impl/mls.hpp

```diff
@@ -616,37 +480,34 @@ pcl::MovingLeastSquares<PointInT, PointOutT>::performUpsampling (PointCloudOut &
         continue;
 
       Eigen::Vector3d add_point = distinct_cloud_->points[dp_i].getVector3fMap ().template cast<double> ();
+      double u, v, w;
+      getMLSCoordinates (add_point, mls_results_[input_index], u, v, w);
 
-      float u_disp = static_cast<float> ((add_point - mls_results_[input_index].mean).dot (mls_results_[input_index].u_axis)),
-            v_disp = static_cast<float> ((add_point - mls_results_[input_index].mean).dot (mls_results_[input_index].v_axis));
-
-      PointOutT result_point;
-      pcl::Normal result_normal;
-      projectPointToMLSSurface (u_disp, v_disp,
-                                mls_results_[input_index].u_axis, mls_results_[input_index].v_axis,
-                                mls_results_[input_index].plane_normal,
-                                mls_results_[input_index].mean,
-                                mls_results_[input_index].curvature,
-                                mls_results_[input_index].c_vec,
-                                mls_results_[input_index].num_neighbors,
-                                result_point, result_normal);
-
-      // Copy additional point information if available
-      copyMissingFields (input_->points[input_index], result_point);
-
-      // Store the id of the original point
-      corresponding_input_indices_->indices.push_back (input_index);
-
-      output.push_back (result_point);
-      if (compute_normals_)
-        normals_->push_back (result_normal);
+      MLSProjectionResults proj;
+      if (polynomial_fit_ && mls_results_[input_index].num_neighbors >= 5 * nr_coeff_)
+      {
+        if (projection_method_ == SIMPLE)
+          proj = projectPointSimpleToPolynomialSurface (u, v, mls_results_[input_index]);
+        else if (projection_method_ == ORTHOGONAL)
+          proj = projectPointOrthogonalToPolynomialSurface (u, v, w, mls_results_[input_index]);
+        else
```

Comment:
You can tweak your if condition to avoid having this internal else branch and use simply the outer one